OpenID Connect groups claim

ijansky opened the pull request:
## Description

The OpenID Connect connector currently does not implement the `GetGroups` method. Since OIDC is used by ADFS, Azure, Google, and Keycloak IPDs, these IDPs currently do not support group based access. This PR attempts to address this.

Addresses CSS-1618

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests
- [ ] Independent change*

## Test instructions

Set up an ADFS/Azure/Google/Keycloak IDP with user groups and verify that group membership defined in the IDP is correctly propagated to and handled by Candid.

## Notes for code reviewers

It may be necessary to handle a case when the IDP sends a response with a URL to fetch groups in another call, instead of a list of all groups. See the [Groups overage claim](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/master/5-WebApp-AuthZ/5-2-Groups/README.md#the-groups-overage-claim) section for details.

It also needs to be determined, whether the "register" flow needs to be altered to support groups.

mhilton commented on internal/candidtest/discharge.go:
```diff
@@ -61,7 +66,7 @@ func (s *DischargeCreator) AssertDischarge(c *qt.C, i httpbakery.Interactor) {
 // macaroon slice containing a discharged macaroon or any error. The
 // newly minted macaroon will have a third-party caveat addressed to the
-// identity server with the given condition.
+// identity server which has the given condition.
```
I wasn't I just lost part of the old one in my changes and didn't remember it right. I'll just change it back.

mhilton replied:
done

mhilton commented on idp/ldap/discharge_test.go:
```diff
@@ -0,0 +1,38 @@
+// Copyright 2015 Canonical Ltd.
```
done

mhilton commented on internal/candidtest/discharge.go:
```diff
@@ -101,3 +106,77 @@ type VisitorFunc func(*url.URL) error
+// OpenWebBrowser returns a function that simulates opening a web browser
+// to complete a login. This function only returns a non-nil error to
+// its callerif there is an error initialising the client. If rh
+// is non-nil it will be called with *http.Response that was received by
```
done

mhilton commented on internal/candidtest/discharge.go:
```diff
@@ -101,3 +106,77 @@ type VisitorFunc func(*url.URL) error
+
+// OpenWebBrowser returns a function that simulates opening a web browser
+// to complete a login. This function only returns a non-nil error to
+// its callerif there is an error initialising the client. If rh
```
done

mhilton commented on internal/candidtest/discharge.go:
```diff
@@ -101,3 +106,77 @@ type VisitorFunc func(*url.URL) error
+			}
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode >= http.StatusBadRequest {
```
done

mhilton commented on internal/candidtest/discharge.go:
```diff
@@ -101,3 +106,77 @@ type VisitorFunc func(*url.URL) error
+		if err != nil {
+			return nil, errgo.Mask(err, errgo.Any)
+		}
+		parts := bytes.Split(buf, []byte("\n"))
```
done

ijansky-canonical committed:
OpenID Connect groups claim

alesstimec reviewed (approved):
lgtm

alesstimec committed:
Merge branch 'master' into oidc_groups

ijansky-canonical committed:
Extensible OIDC groups claim handler

OpenID Connect IDPs may implement their specific handler to retrieve user groups, such as ADFS or Azure with Microsoft Graph API.

alnvdl commented on idp/openid/openid-connect.go:
```diff
@@ -66,6 +68,13 @@ type IdentityCreator interface {
+// A GroupsRetriever is used to retrieve a list of user groups from the
+// OpenID token returned by the OpenID authentication process.
+type GroupsRetriever interface {
+	// Retrieve groups from the OpenID token.
```
It's a bit silly, but usually godocs of functions are written with the name of the function as the first word: "RetrieveGroups retrieves groups from the OpenID token".

alnvdl commented on idp/openid/openid-connect.go:
```diff
@@ -397,16 +418,31 @@ func (idp *openidConnectIdentityProvider) CreateIdentity(ctx context.Context, to
 		}
 		user.Email = claims.Email
 		user.Name = claims.FullName
+		user.Groups, err = idp.RetrieveGroups(ctx, tok, &claims, id.Claims)
```
`id.Claims(&claims)` was called above. Why does `id.Claims` then gets passed to `RetrieveGroups` for calling it again and checking the error?

Technically, and because of this, I guess it would be impossible to reach that "Failed to unmarshal claims" error in `msgraph.go`.

alnvdl commented on idp/idputil/msgraph.go:
```diff
@@ -0,0 +1,116 @@
+	var ok bool
+
+	var claimName string
+	if claimName, ok = claims.ClaimsNames["groups"]; !ok {
```
No need to declare `ok` and `claimName` above. They can be declared inline here with:
```go
if claimName, ok := claims.ClaimsNames["groups"]; !ok {
...
```

alnvdl commented on idp/idputil/msgraph.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	var claimSource claimSource
+	if claimSource, ok = claims.ClaimsSources[claimName]; !ok {
```
Same here.

alnvdl commented on idp/idputil/msgraph.go:
```diff
@@ -0,0 +1,116 @@
+	req.Header.Set("Authorization", "Bearer "+token.AccessToken)
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
```
It'd be good to check with @alesstimec if this project usually uses the default HTTP client or instead uses a custom instantiation. For unit testing purposes, it's usually a good idea for each package to be in control of their own HTTP clients, but it's not a hard rule.

alnvdl commented on idp/idputil/msgraph.go:
```diff
@@ -0,0 +1,116 @@
+// Copyright 2015 Canonical Ltd.
```
2022?

alnvdl reviewed (approved):
LGTM, with a few suggestions.

ijansky-canonical committed:
Extensible OIDC groups claim handler - code review

ijansky-canonical committed:
Extensible OIDC groups claim handler - registration form

alesstimec commented on idp/adfs/adfs.go:
```diff
@@ -78,16 +79,21 @@ func NewIdentityProvider(p Params) idp.IdentityProvider {
+		ClientSecret:    p.ClientSecret,
+		Hidden:          p.Hidden,
+		MatchEmailAddr:  p.MatchEmailAddr,
+		GroupsRetriever: &groupsRetriever,
```
GroupsRetriever: &msgraph.MsGraphGroupsRetriever{},

alesstimec commented on idp/azure/azure.go:
```diff
@@ -70,15 +71,18 @@ func NewIdentityProvider(p Params) idp.IdentityProvider {
+		ClientID:        p.ClientID,
+		ClientSecret:    p.ClientSecret,
+		Hidden:          p.Hidden,
+		GroupsRetriever: &groupsRetriever,
```
GroupsRetriever: &msgraph.MsGraphGroupsRetriever{},

alesstimec commented on idp/idputil/idputil.go:
```diff
@@ -259,3 +265,42 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+// ReadGroupsFromCSV parses a CSV string with user groups
+// and returns groups as a slice
+func ReadGroupsFromCSV(src string) ([]string, error) {
+	r := csv.NewReader(strings.NewReader(src))
```
you could just use strings.Split(src, ",")

alesstimec commented on idp/idputil/idputil.go:
```diff
@@ -113,6 +115,10 @@ type RegistrationParams struct {
 	// populate the email input.
 	Email string
+
+	// Groups contains a CSV formatted list of groups the user is
```
contains a comma separated list of groups the user  is a member of.

alesstimec commented on idp/idputil/idputil.go:
```diff
@@ -259,3 +265,42 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+
+// WriteGroupsToCSV writes a slice of user groups to
+// a CSV string
+func WriteGroupsToCSV(groups []string) (string, error) {
```
you could use strings.Join(groups, ",")

alesstimec reviewed (commented):
looks good and works. I'd suggest simplification of how groups are written and parsed from comma separated strings.. 

also just a note: this will record group IDs not, the actual groups names. Is that ok?

ijansky-canonical committed:
Extensible OIDC groups claim handler - code review

alesstimec committed:
Marshal user groups as JSON instead of CSV.

ijansky committed:
Merge pull request #1 from alesstimec/oidc-groups-json

Marshal user groups as JSON instead of CSV.
